File: mro/stages/cnv/get_uniq_coverage_bedGraph/binning.py
# Copyright (c) 2017 10X Genomics, Inc. All rights reserved.
import math


def Harmonic(x):
    return 2.0 / 3.0 * math.pow(x, 1.5) + math.sqrt(x) / 2.0 - 0.2


def InvHarmonic(y):
    return math.pow(1.5 * (y + 0.2), 2.0 / 3.0)


def getBins(n, amp):
    numBins = int(round(InvHarmonic(n * 1.0 / amp))) + 1
    bins = [int(round(Harmonic(x) * amp)) for x in range(numBins)]
    return bins


def discretize(num, bins, amp):
    idx = int(round(InvHarmonic(num * 1.0 / amp)))
    if num > bins[idx]:
        print "Wrong! idx in discreteize function is too low", num, bins[idx]
        return -1
    while idx > 0 and num <= bins[idx - 1]:
        idx = idx - 1
    return idx

File: lib/go/src/cnv/wgs/binning.go
// Copyright (c) 2017 10X Genomics, Inc. All rights reserved.
package main

// recommended amp 1.5

import (
	"fmt"
	"math"
)

func GetBins(n int, amp float64) (bins []int) {
	// n is the largest possible read count
	numBins := int(math.Floor(InvHarmonic(float64(n)/amp)+0.5)) + 1
	bins = make([]int, numBins)
	for i := 0; i < numBins; i++ {
		bins[i] = int(math.Floor(Harmonic(float64(i))*amp + 0.5))
		fmt.Println(i, Harmonic(float64(i))*amp)
	}
	return bins
}

func Harmonic(x float64) float64 {
	return 0.66666667*math.Pow(x, 1.5) + math.Sqrt(x)/2.0 - 0.2
}

func RHarmonic(x float64) int {
	return int(math.Floor(Harmonic(x) + 0.5))
}

func InvHarmonic(x float64) float64 {
	return math.Pow(1.5*x, 2.0/3.0)
}
